patina_dxe_core/src/pi_dispatcher/fv.rs

```diff
@@ -0,0 +1,1705 @@
+//! DXE Core Firmware Volume (FV)
+//!
+//! ## License
+//!
+//! Copyright (c) Microsoft Corporation.
+//!
+//! SPDX-License-Identifier: Apache-2.0
+//!
+use core::{
+    ffi::c_void,
+    mem::{self, size_of},
+    num::NonZeroUsize,
+    ptr::NonNull,
+    slice,
+};
+
+use alloc::{boxed::Box, collections::BTreeMap};
+use patina::pi::{
+    self,
+    fw_fs::{ffs, fv, fvb},
+    hob,
+};
+
+use patina::error::EfiError;
+use patina_ffs::{file::FileRef, section::SectionExtractor, volume::VolumeRef};
+use patina_internal_device_path::concat_device_path_to_boxed_slice;
+use r_efi::efi;
+
+use crate::{
+    Core, PlatformInfo,
+    allocator::core_allocate_pool,
+    protocols::{PROTOCOL_DB, core_install_protocol_interface},
+    tpl_mutex,
+};
+
+/// A container for a FV or FVB protocol instance.
+///
+/// The protocol instances themselves are not used directly by rust code, but this container is used to manage the
+/// lifetime of the protocols and to validate accesses to them.
+#[allow(dead_code)]
+enum Protocol {
+    /// A heap-allocated FV protocol instance.
+    Fv(Box<pi::protocols::firmware_volume::Protocol>),
+    /// A heap-allocated FVB protocol instance.
+    Fvb(Box<pi::protocols::firmware_volume_block::Protocol>),
+}
+
+/// The metadata associated with a given FV / FVB protocol installation.
+struct Metadata {
+    /// The installed protocol instance.
+    protocol: Protocol,
+    /// The physical address of the FV / FVB associated with the protocol.
+    physical_address: u64,
+}
+
+impl Metadata {
+    /// Creates a new Metadata instance for a FVB protocol.
+    fn new_fvb(protocol: Box<pi::protocols::firmware_volume_block::Protocol>, physical_address: u64) -> Self {
+        Self { protocol: Protocol::Fvb(protocol), physical_address }
+    }
+
+    /// Creates a new Metadata instance for a FV protocol.
+    fn new_fv(protocol: Box<pi::protocols::firmware_volume::Protocol>, physical_address: u64) -> Self {
+        Self { protocol: Protocol::Fv(protocol), physical_address }
+    }
+}
+
+/// Stored protocol data for any FV/FVB protocols installed by the DXE core.
+pub(super) struct FvProtocolData<P: PlatformInfo> {
+    /// A map of installed FV/FVB protocol pointers (key) and the corresponding metadata (value).
+    fv_metadata: BTreeMap<NonZeroUsize, Metadata>,
+    /// A marker for accessing the singleton `Core` instance in a UEFI protocol method.
+    _platform_info: core::marker::PhantomData<P>,
+}
+
+impl<P: PlatformInfo> FvProtocolData<P> {
+    /// Returns the FV's physical address for the given protocol pointer, if it is in-fact a FV protocol.
+    #[inline(always)]
+    fn get_fv_address(&self, protocol: NonNull<pi::protocols::firmware_volume::Protocol>) -> Option<u64> {
```

Comment:
For my education, what is the reason for `get_fv_address` and `get_fvb_address` if they return the same thing?